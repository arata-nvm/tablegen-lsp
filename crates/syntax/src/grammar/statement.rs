use crate::{
    grammar::{delimited, r#type, value},
    parser::Parser,
    syntax_kind::SyntaxKind,
    token_kind::TokenKind,
    T,
};

pub(super) enum StatementListType {
    TopLevel,
    Block,
    SingleOrBlock,
}

// StatementList ::= Statement*
pub(super) fn statement_list(p: &mut Parser, typ: StatementListType) {
    p.start_node(SyntaxKind::StatementList);
    p.skip();
    match typ {
        StatementListType::TopLevel => {
            while !p.eof() {
                statement(p);
            }
        }
        StatementListType::Block => {
            p.expect(T!['{']);
            while !p.at(T!['}']) && !p.eof() {
                statement(p);
            }
            p.expect(T!['}']);
        }
        StatementListType::SingleOrBlock => {
            if p.eat_if(T!['{']) {
                while !p.at(T!['}']) && !p.eof() {
                    statement(p);
                }
                p.expect(T!['}']);
            } else {
                statement(p);
            }
        }
    }
    p.finish_node();
}

// Statement ::= Include | Assert | Class | Def | Defm | Defset | Defvar | Foreach | If | Let | MultiClass
pub(super) fn statement(p: &mut Parser) {
    match p.peek() {
        T![include] => include(p),
        T![assert] => r#assert(p),
        T![class] => class(p),
        T![def] => def(p),
        T![defm] => defm(p),
        T![defset] => defset(p),
        T![defvar] => defvar(p),
        T![foreach] => foreach(p),
        T![if] => r#if(p),
        T![let] => r#let(p),
        T![multiclass] => multi_class(p),
        _ => p.error_and_eat("expected class, def, defm, defset, multiclass, let or foreach"),
    }
}

// Include ::= "include" String
pub(super) fn include(p: &mut Parser) {
    p.start_node(SyntaxKind::Include);
    p.assert(T![include]);
    value::string(p).or_error(p, "expected filename after include");
    p.finish_node();
}

// Class ::= "class" Identifier TemplateArgList? RecordBody
pub(super) fn class(p: &mut Parser) {
    p.start_node(SyntaxKind::Class);
    p.assert(T![class]);
    value::identifier(p).or_error(p, "expected class name after 'class' keyword");
    opt_template_arg_list(p);
    record_body(p);
    p.finish_node();
}

// Def ::= "def" Value? RecordBody
pub(super) fn def(p: &mut Parser) {
    p.start_node(SyntaxKind::Def);
    p.assert(T![def]);
    object_name(p);
    record_body(p);
    p.finish_node();
}

pub(super) fn object_name(p: &mut Parser) {
    match p.peek() {
        T![:] | T![;] | T!['{'] => {}
        _ => value::opt_name_value(p),
    }
}

// Let ::= "let" LetList "in" ( "{" Statement* "}" | Statement );
pub(super) fn r#let(p: &mut Parser) {
    p.start_node(SyntaxKind::Let);
    p.assert(T![let]);
    let_list(p);
    p.expect_with_msg(T![in], "expected 'in' at end of top-level 'let'");
    statement_list(p, StatementListType::SingleOrBlock);
    p.finish_node();
}

// LetList ::= LetItem ( "," LetItem )*
pub(super) fn let_list(p: &mut Parser) {
    p.start_node(SyntaxKind::LetList);
    while !p.eof() {
        let_item(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
}

// LetItem ::= Identifier ( "<" RangeList ">" )? "=" Value
pub(super) fn let_item(p: &mut Parser) {
    p.start_node(SyntaxKind::LetItem);
    value::identifier(p).or_error(p, "expected identifier in let expression");
    if p.eat_if(T![<]) {
        value::range_list(p);
        p.expect_with_msg(T![>], "expected '>' at end of range list");
    }
    p.expect_with_msg(T![=], "expected '=' in let expression");
    value::value(p);
    p.finish_node();
}

// MultiClass ::= "multiclass" Identifier TemplateArgList? ParentClassList "{" MultiClassStatement+ "}"
pub(super) fn multi_class(p: &mut Parser) {
    p.start_node(SyntaxKind::MultiClass);
    p.assert(T![multiclass]);
    value::identifier(p).or_error(p, "expected identifier after multiclass for name");
    opt_template_arg_list(p);
    parent_class_list(p);
    p.expect_with_msg(T!['{'], "expected '{' in multiclass definition");
    multi_class_statements(p);
    p.finish_node();
}

pub(super) fn multi_class_statements(p: &mut Parser) {
    p.start_node(SyntaxKind::StatementList);
    multi_class_statement(p);
    while !p.at(T!['}']) && !p.eof() {
        multi_class_statement(p);
    }
    p.expect(T!['}']);
    p.finish_node();
}

// MultiClassStatement ::= Def | Defm | Foreach | Let | If
pub(super) fn multi_class_statement(p: &mut Parser) {
    match p.peek() {
        T![def] => def(p),
        T![defm] => defm(p),
        T![foreach] => foreach(p),
        T![let] => r#let(p),
        T![assert] => r#assert(p),
        T![if] => r#if(p),
        _ => p.error_and_eat("expected 'let', 'def', 'defm' or 'foreach' in multiclass body"),
    }
}

// Defm ::= "defm" Value? ParentClassList ";"
pub(super) fn defm(p: &mut Parser) {
    p.start_node(SyntaxKind::Defm);
    p.assert(T![defm]);
    object_name(p);
    parent_class_list(p);
    p.expect_with_msg(T![;], "expected ';' at end of defm");
    p.finish_node();
}

// Defset ::= "defset" Type Identifier "=" "{" Statement* "}"
pub(super) fn defset(p: &mut Parser) {
    p.start_node(SyntaxKind::Defset);
    p.assert(T![defset]);
    r#type::r#type(p);
    value::identifier(p);
    p.expect(T![=]);
    statement_list(p, StatementListType::Block);
    p.finish_node();
}

// Defvar ::= "defvar" Identifier "=" Value ";"
pub(super) fn defvar(p: &mut Parser) {
    p.start_node(SyntaxKind::Defvar);
    p.assert(T![defvar]);
    value::identifier(p);
    p.expect(T![=]);
    value::value(p);
    p.expect(T![;]);
    p.finish_node();
}

// Foreach ::= "foreach" ForeachIterator "in" ( "{" Statement* "}" | Statement )
pub(super) fn foreach(p: &mut Parser) {
    p.start_node(SyntaxKind::Foreach);
    p.assert(T![foreach]);
    foreach_iterator(p);
    p.expect(T![in]);
    statement_list(p, StatementListType::SingleOrBlock);
    p.finish_node();
}

// ForeachIterator ::= Identifier "=" ForeachIteratorInit
pub(super) fn foreach_iterator(p: &mut Parser) {
    p.start_node(SyntaxKind::ForeachIterator);
    value::identifier(p).or_error(p, "expected identifier in foreach declaration");
    p.expect_with_msg(T![=], "expected '=' in foreach declaration");
    foreach_iterator_init(p);
    p.finish_node()
}

// ForeachIteratorInit ::= "{" RangeList "}" | RangePiece | Value
pub(super) fn foreach_iterator_init(p: &mut Parser) {
    match p.peek() {
        T!['{'] => {
            p.assert(T!['{']);
            let m = value::range_list(p);
            p.expect_with_msg(T!['}'], "expected '}' at end of bit range list");
            m
        }
        TokenKind::IntVal => value::range_piece(p),
        _ => value::value(p),
    };
}

// If ::= "if" Value "then" ( "{" Statement* "}" | Statement ) ( "else" ( "{" Statement* "}" | Statement ) )?
fn r#if(p: &mut Parser) {
    p.start_node(SyntaxKind::If);
    p.assert(T![if]);
    value::value(p);
    p.expect(T![then]);
    statement_list(p, StatementListType::SingleOrBlock);
    if p.eat_if(T![else]) {
        statement_list(p, StatementListType::SingleOrBlock);
    }
    p.finish_node();
}

// Assert ::= "assert" Value "," Value ";"
fn r#assert(p: &mut Parser) {
    p.start_node(SyntaxKind::Assert);
    p.assert(T![assert]);
    value::value(p);
    p.expect(T![,]);
    value::value(p);
    p.expect(T![;]);
    p.finish_node();
}

pub(super) fn opt_template_arg_list(p: &mut Parser) {
    if p.at(T![<]) {
        template_arg_list(p);
    }
}

// TemplateArgList ::= "<" TemplateArgDecl ( "," TemplateArgDecl )* ">"
pub(super) fn template_arg_list(p: &mut Parser) {
    p.start_node(SyntaxKind::TemplateArgList);
    delimited(p, T![<], T![>], T![,], template_arg_decl);
    p.finish_node();
}

// TemplateArgDecl ::= Type Identifier ( "=" Value )?
pub(super) fn template_arg_decl(p: &mut Parser) {
    p.start_node(SyntaxKind::TemplateArgDecl);
    r#type::r#type(p);
    value::identifier(p).or_error(p, "expected identifier in declaration");
    if p.eat_if(T![=]) {
        value::value(p);
    }
    p.finish_node();
}

// RecordBody ::= ParentClassList Body
pub(super) fn record_body(p: &mut Parser) {
    p.start_node(SyntaxKind::RecordBody);
    parent_class_list(p);
    body(p);
    p.finish_node();
}

// ParentClassList ::= ( ":" ClassRef ( "," ClassRef )* )?
pub(super) fn parent_class_list(p: &mut Parser) {
    p.start_node(SyntaxKind::ParentClassList);
    if !p.eat_if(T![:]) {
        p.finish_node();
        return;
    }

    while !p.eof() {
        class_ref(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
}

// ClassRef ::= Identifier ( "<" ArgValueList? ">" )?
pub(super) fn class_ref(p: &mut Parser) {
    p.start_node(SyntaxKind::ClassRef);
    value::identifier(p);
    if p.eat_if(T![<]) {
        arg_value_list(p);
        p.expect_with_msg(T![>], "expected '>' in template value list");
    }
    p.finish_node();
}

// ArgValueList ::= PositionalArgValueList ","? NamedArgValueList
pub(super) fn arg_value_list(p: &mut Parser) {
    p.start_node(SyntaxKind::ArgValueList);
    if p.at_set(&value::VALUE_START) {
        positional_arg_value_list(p);
    }
    // TODO
    // p.eat_if(T![,]);
    // if p.at_set(&VALUE_START) {
    // named_arg_value_list(p);
    // }
    p.finish_node();
}

// PositionalArgValueList ::= ( Value ( "," Value )* ) ?
pub(super) fn positional_arg_value_list(p: &mut Parser) {
    p.start_node(SyntaxKind::PositionalArgValueList);
    while !p.eof() {
        value::value(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
}

// NamedArgValueList ::= ( NamedArgValue ( "," NamedArgValue )* )?
#[allow(unused)]
pub(super) fn named_arg_value_list(p: &mut Parser) {
    p.start_node(SyntaxKind::NamedArgValueList);
    while !p.eof() {
        named_arg_value(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
}

// NamedArgValue ::= Value "=" Value
#[allow(unused)]
pub(super) fn named_arg_value(p: &mut Parser) {
    p.start_node(SyntaxKind::NamedArgValue);
    value::value(p);
    p.expect(T![=]);
    value::value(p);
    p.finish_node();
}

// Body ::= ";" | "{" BodyItem* "}"
pub(super) fn body(p: &mut Parser) {
    p.start_node(SyntaxKind::Body);
    if p.eat_if(T![;]) {
        p.finish_node();
        return;
    }

    p.expect_with_msg(T!['{'], "expected ';' or '{' to start body");
    while !p.at(T!['}']) && !p.eof() {
        if !body_item(p) {
            break;
        }
    }
    p.expect(T!['}']);
    p.finish_node();
}

// BodyItem ::= FieldDef | FieldLet | Defvar | Assert
pub(super) fn body_item(p: &mut Parser) -> bool {
    if p.at_set(&r#type::TYPE_FIRST_TOKENS) || p.at(T![field]) {
        field_def(p);
        return true;
    }

    match p.peek() {
        T![let] => field_let(p),
        T![defvar] => defvar(p),
        T![assert] => r#assert(p),
        _ => return false,
    }

    true
}

// FieldDef ::= "field"? ( Type | CodeType ) Identifier ( "=" Value )? ";"
pub(super) fn field_def(p: &mut Parser) {
    p.start_node(SyntaxKind::FieldDef);
    p.eat_if(T![field]);
    r#type::r#type(p);
    value::identifier(p).or_error(p, "expected identifier in declaration");
    if p.eat_if(T![=]) {
        value::value(p);
    }
    p.expect_with_msg(T![;], "expected ';' after declaration");
    p.finish_node();
}

// FieldLet ::= "let" Identitfer ( "{" RangeList "}" )? "=" Value ";"
pub(super) fn field_let(p: &mut Parser) {
    p.start_node(SyntaxKind::FieldLet);
    p.assert(T![let]);
    value::identifier(p).or_error(p, "expected field identifier after let");
    if p.eat_if(T!['{']) {
        value::range_list(p);
        p.expect_with_msg(T!['}'], "expected '}' at end of bit list");
    }
    p.expect(T![=]);
    value::value(p).or_error(p, "expected '=' in let expression");
    p.expect_with_msg(T![;], "expected ';' after let expression");
    p.finish_node();
}
