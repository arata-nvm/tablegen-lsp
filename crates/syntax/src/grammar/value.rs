use crate::{
    parser::{CompletedMarker, Parser},
    syntax_kind::SyntaxKind,
    token_kind::TokenKind,
    T,
};

use super::{delimited, r#type, statement};

pub(super) const VALUE_START: [TokenKind; 63] = [
    TokenKind::IntVal,
    TokenKind::BinaryIntVal,
    TokenKind::StrVal,
    TokenKind::CodeFragment,
    T![true],
    T![false],
    T![?],
    T!['{'],
    T!['['],
    T!['('],
    TokenKind::Id,
    TokenKind::VarName,
    T![!con],
    T![!add],
    T![!sub],
    T![!mul],
    T![!div],
    T![!not],
    T![!log2],
    T![!and],
    T![!or],
    T![!xor],
    T![!sra],
    T![!srl],
    T![!shl],
    T![!listconcat],
    T![!listsplat],
    T![!strconcat],
    T![!initialized],
    T![!interleave],
    T![!substr],
    T![!find],
    T![!cast],
    T![!subst],
    T![!foreach],
    T![!filter],
    T![!foldl],
    T![!head],
    T![!tail],
    T![!size],
    T![!empty],
    T![!if],
    T![!cond],
    T![!eq],
    T![!isa],
    T![!dag],
    T![!ne],
    T![!le],
    T![!lt],
    T![!ge],
    T![!gt],
    T![!setdagop],
    T![!getdagop],
    T![!exists],
    T![!listremove],
    T![!tolower],
    T![!toupper],
    T![!range],
    T![!getdagarg],
    T![!getdagname],
    T![!setdagarg],
    T![!setdagname],
    T![!repr],
];

pub(super) fn opt_value(p: &mut Parser) {
    if p.at_set(&VALUE_START) {
        value(p);
    }
}

// Value ::= InnerValue ( "#" InnerValue )*
pub(super) fn value(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Value);
    inner_value(p);
    while p.eat_if(T![#]) {
        inner_value(p);
    }
    p.finish_node();
    CompletedMarker::Success
}

// InnerValue ::= SimpleValue ValueSuffix*
pub(super) fn inner_value(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::InnerValue);
    if !simple_value(p).is_success() {
        p.finish_node();
        return CompletedMarker::Fail;
    }

    while value_suffix(p) {}

    p.finish_node();
    CompletedMarker::Success
}

pub(super) fn opt_name_value(p: &mut Parser) {
    if p.at_set(&VALUE_START) {
        name_value(p);
    }
}

// Value(NameMode) ::= InnerValue ( "#" InnerValue )*
pub(super) fn name_value(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Value);
    inner_name_value(p);
    while p.eat_if(T![#]) {
        inner_name_value(p);
    }
    p.finish_node();
    CompletedMarker::Success
}

// InnerValue(NameMode) ::= SimpleValue ValueSuffix*
pub(super) fn inner_name_value(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::InnerValue);
    if !simple_value(p).is_success() {
        p.finish_node();
        return CompletedMarker::Fail;
    }

    while !p.at(T!['{']) && value_suffix(p) {}

    p.finish_node();
    CompletedMarker::Success
}

// ValueSuffix ::= RangeSuffix | SliceSuffix | FieldSuffix
pub(super) fn value_suffix(p: &mut Parser) -> bool {
    match p.peek() {
        T!['{'] => range_suffix(p),
        T!['['] => slice_suffix(p),
        T![.] => field_suffix(p),
        _ => return false,
    };

    true
}

// RangeSuffix ::= "{" RangeList "}"
pub(super) fn range_suffix(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::RangeSuffix);
    p.assert(T!['{']);
    range_list(p);
    p.expect_with_msg(T!['}'], "expected '}' at end of bit range list");
    p.finish_node();
    CompletedMarker::Success
}

// RangeList ::= RangePiece ( "," RangePiece )*
pub(super) fn range_list(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::RangeList);
    while !p.eof() {
        range_piece(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
    CompletedMarker::Success
}

// RangePiece ::= Integer | Integer "..." Integer | Integer "-" Integer | Integer Integer
pub(super) fn range_piece(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::RangePiece);
    integer(p).or_error(p, "expected integer or bitrange");
    if p.at_set(&[T![...], T![-]]) {
        p.eat();
    }
    if p.at(TokenKind::IntVal) {
        integer(p).or_error(p, "expected integer value as end of range");
    }
    p.finish_node();
    CompletedMarker::Success
}

// SliceSuffix ::= "[" SliceElements "]"
pub(super) fn slice_suffix(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::SliceSuffix);
    p.assert(T!['[']);
    slice_elements(p);
    p.expect_with_msg(T![']'], "expected ']' at end of list slice");
    p.finish_node();
    CompletedMarker::Success
}

// SliceElements ::= ( SliceElement "," )* SliceElement ","?
pub(super) fn slice_elements(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::SliceElements);
    while !p.eof() {
        slice_element(p);
        if !p.eat_if(T![,]) || p.at(T![']']) {
            break;
        }
    }
    p.finish_node();
    CompletedMarker::Success
}

// SliceElement ::= Value | Value "..." Value | Value "-" Value | Value Integer
pub(super) fn slice_element(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::SliceElement);
    value(p);
    if p.at_set(&[T![...], T![-]]) {
        p.eat();
    }
    opt_value(p);
    p.finish_node();
    CompletedMarker::Success
}

// FieldSuffix ::= "." Identifier
pub(super) fn field_suffix(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::FieldSuffix);
    p.assert(T![.]);
    identifier(p).or_error(p, "expected field identifier after '.'");
    p.finish_node();
    CompletedMarker::Success
}

// SimpleValue ::= Integer | String | Code | Boolean | Uninitialized | Bits | List | Dag | Identifier | ClassValue | BangOperator | CondOperator
pub(super) fn simple_value(p: &mut Parser) -> CompletedMarker {
    match p.peek() {
        TokenKind::IntVal | TokenKind::BinaryIntVal => integer(p),
        TokenKind::StrVal => string(p),
        TokenKind::CodeFragment => code(p),
        T![true] | T![false] => boolean(p),
        T![?] => uninitialized(p),
        T!['{'] => bits(p),
        T!['['] => list(p),
        T!['('] => dag(p),
        TokenKind::Id => identifier_or_class_value(p),
        kind if kind.is_bang_operator() => bang_operator(p),
        kind if kind.is_cond_operator() => cond_operator(p),
        _ => {
            p.error_and_recover("unknown token when parsing a value");
            CompletedMarker::Fail
        }
    }
}

// Integer ::= INT
pub(super) fn integer(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Integer);
    if p.eat_if(TokenKind::IntVal) || p.eat_if(TokenKind::BinaryIntVal) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// String ::= STRING
pub(super) fn string(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::String);
    if p.at(TokenKind::StrVal) {
        while p.eat_if(TokenKind::StrVal) {}
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// Code ::= CODE
pub(super) fn code(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Code);
    if p.eat_if(TokenKind::CodeFragment) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// Boolean ::= "true" | "false"
pub(super) fn boolean(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Boolean);
    if p.eat_if(T![true]) || p.eat_if(T![false]) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// Uninitialized ::= "?"
pub(super) fn uninitialized(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Uninitialized);
    if p.eat_if(T![?]) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// Bits ::= "{" ValueList "}"
pub(super) fn bits(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Bits);
    value_list(p, T!['{'], T!['}']);
    p.finish_node();
    CompletedMarker::Success
}

// List ::= "[" ValueList "]"
pub(super) fn list(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::List);
    value_list(p, T!['['], T![']']);
    if p.eat_if(T![<]) {
        r#type::r#type(p);
        p.expect_with_msg(T![>], "expected '>' at end of list element type");
    }
    p.finish_node();
    CompletedMarker::Success
}

// ValueList ::= Value ( "," Value )*
pub(super) fn value_list(p: &mut Parser, bra: TokenKind, ket: TokenKind) -> CompletedMarker {
    p.start_node(SyntaxKind::ValueList);
    delimited(p, bra, ket, T![,], |p| {
        value(p);
    });
    p.finish_node();
    CompletedMarker::Success
}

// Dag ::= ( DagArg DagArgList? )
pub(super) fn dag(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Dag);
    p.expect(T!['(']);
    if !p.at_set(&[TokenKind::Id, T![!cast], T![?], T![!getdagop]]) {
        p.error("expected identifier in dag init");
        p.finish_node();
        return CompletedMarker::Fail;
    }
    dagarg(p);
    if !p.at(T![')']) {
        dagarg_list(p);
    }
    p.expect_with_msg(T![')'], "expected ')' in dag init");
    p.finish_node();
    CompletedMarker::Success
}

// DagArgList ::= DagArg ( "," DagArg )*
pub(super) fn dagarg_list(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::DagArgList);
    while !p.eof() {
        dagarg(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.finish_node();
    CompletedMarker::Success
}

// DagArg ::= Value ( ":" VARNAME ) | VARNAME
pub(super) fn dagarg(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::DagArg);
    if p.eat_if(TokenKind::VarName) {
        p.finish_node();
        return CompletedMarker::Success;
    }

    value(p);
    if p.eat_if(T![:]) {
        var_name(p).or_error(p, "expected variable name in dag literal");
    }
    p.finish_node();
    CompletedMarker::Success
}

// VarName ::= VARNAME
pub(super) fn var_name(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::VarName);
    if p.eat_if(TokenKind::VarName) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// Identifier ::= ID
pub(super) fn identifier(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::Identifier);
    if p.eat_if(TokenKind::Id) {
        p.finish_node();
        CompletedMarker::Success
    } else {
        p.finish_node();
        CompletedMarker::Fail
    }
}

// ClassValue ::= Identifier ( "<" ArgValueList? ">" )?
pub(super) fn identifier_or_class_value(p: &mut Parser) -> CompletedMarker {
    let c = p.builder().checkpoint();
    let m = identifier(p);
    if p.eat_if(T![<]) {
        p.builder().start_node_at(c, SyntaxKind::ClassValue.into());
        statement::arg_value_list(p);
        p.expect_with_msg(T![>], "expected '>' at end of value list");
        p.finish_node();
        CompletedMarker::Success
    } else {
        m
    }
}

// BangOperator ::= BANGOP ( "<" Type ">" )? "(" ValueList ")"
pub(super) fn bang_operator(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::BangOperator);
    if !p.peek().is_bang_operator() {
        p.error_and_recover("expected bang operator");
        p.finish_node();
        return CompletedMarker::Fail;
    }
    p.eat(); // eat bang operator
    if p.eat_if(T![<]) {
        r#type::r#type(p);
        p.expect(T![>]);
    }
    delimited(p, T!['('], T![')'], T![,], |p| {
        value(p);
    });
    p.finish_node();
    CompletedMarker::Success
}

// CondOperator ::= CONDOP "(" CondClause ( "," CondClause )* ")"
pub(super) fn cond_operator(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::CondOperator);
    p.expect(T![!cond]);
    delimited(p, T!['('], T![')'], T![,], |p| {
        cond_clause(p);
    });
    p.finish_node();
    CompletedMarker::Success
}

// CondClause ::= Value ":" Value
pub(super) fn cond_clause(p: &mut Parser) -> CompletedMarker {
    p.start_node(SyntaxKind::CondClause);
    value(p);
    p.expect(T![:]);
    value(p);
    p.finish_node();
    CompletedMarker::Success
}
