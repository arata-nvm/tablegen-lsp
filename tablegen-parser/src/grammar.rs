pub mod statement;
pub mod r#type;
pub mod value;

use crate::{
    error::SyntaxError, grammar::statement::StatementListType, language::SyntaxNode,
    parser::Parser, syntax_kind::SyntaxKind, token_kind::TokenKind, T,
};

const RECOVER_TOKENS: [TokenKind; 5] = [T![include], T![class], T![def], T![let], T![;]];

pub fn parse(text: &str) -> (SyntaxNode, Vec<SyntaxError>) {
    let mut parser = Parser::new(text, &RECOVER_TOKENS);
    root(&mut parser);

    let errors = parser.take_errors();
    (parser.finish(), errors)
}

// Root ::= StatementList
fn root(p: &mut Parser) {
    p.start_node(SyntaxKind::Root);
    statement::statement_list(p, StatementListType::TopLevel);
    if !p.eof() {
        p.error("unexpected input at top level");
    }
    p.finish_node();
}

fn delimited<F>(p: &mut Parser, bra: TokenKind, ket: TokenKind, delim: TokenKind, mut parser: F)
where
    F: FnMut(&mut Parser<'_>),
{
    p.expect(bra);
    while !p.at(ket) && !p.eof() {
        parser(p);

        if !p.eat_if(delim) {
            break;
        }
    }
    p.expect(ket);
}

#[cfg(test)]
mod tests {
    use crate::grammar::parse;

    #[test]
    fn statement() {
        insta::assert_debug_snapshot!(parse("a"));
    }

    #[test]
    fn include() {
        insta::assert_debug_snapshot!(parse(r#"include "foo.td""#));

        insta::assert_debug_snapshot!(parse("include"));
    }

    #[test]
    fn class() {
        insta::assert_debug_snapshot!(parse("class Foo<int A, int B = 1>: Bar<A, 2>;"));
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A> {
                int B;
                int C = A;
                let D = A;
            }"
        ));
        insta::assert_debug_snapshot!(parse("class"));
    }

    #[test]
    fn def() {
        insta::assert_debug_snapshot!(parse("def Foo : Bar;"));
        insta::assert_debug_snapshot!(parse("def foo {}"));
    }

    #[test]
    fn r#let() {
        insta::assert_debug_snapshot!(parse("let A = 1, B<1...3> = 0b101 in { class Foo; }"));

        insta::assert_debug_snapshot!(parse("let A = 1"));
        insta::assert_debug_snapshot!(parse("let A = 1 {"));
    }

    #[test]
    fn let_item() {
        insta::assert_debug_snapshot!(parse("let"));
        insta::assert_debug_snapshot!(parse("let A"));
    }

    #[test]
    fn multi_class() {
        insta::assert_debug_snapshot!(parse(
            "multiclass foo {
                def _foo1;
                def _foo2;
            }"
        ));
    }

    #[test]
    fn defm() {
        insta::assert_debug_snapshot!(parse("defm foo : bar;"));
    }

    #[test]
    fn defset() {
        insta::assert_debug_snapshot!(parse(
            "defset list<Base> BaseList = {
                def Foo0 : Base<0>;
                def Foo1 : Base<1>;
            }"
        ));
    }

    #[test]
    fn defvar() {
        insta::assert_debug_snapshot!(parse("defvar i = 0;"));
    }

    #[test]
    fn foreach() {
        insta::assert_debug_snapshot!(parse(
            "foreach i = [0, 1] in {
                def Foo # i : Base<i>;
            }"
        ));
    }

    #[test]
    fn r#if() {
        insta::assert_debug_snapshot!(parse("if true then { class Foo; }"));
    }

    #[test]
    fn assert() {
        insta::assert_debug_snapshot!(parse("assert hoge, \"fuga\";"));
    }

    #[test]
    fn template_arg_decl() {
        insta::assert_debug_snapshot!(parse("class Foo<int"));
    }

    #[test]
    fn class_ref() {
        insta::assert_debug_snapshot!(parse("class Foo : Bar<"));
    }

    #[test]
    fn r#type() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<bit A, int B, string C, dag D, bits<32> E, list<int> F, Bar G>;"
        ));
    }

    #[test]
    fn body() {
        insta::assert_debug_snapshot!(parse("class Foo"));
    }

    #[test]
    fn field_def() {
        insta::assert_debug_snapshot!(parse("class Foo { int"));
        insta::assert_debug_snapshot!(parse("class Foo { int A"));
    }

    #[test]
    fn field_let() {
        insta::assert_debug_snapshot!(parse("class Foo { let"));
        insta::assert_debug_snapshot!(parse("class Foo { let A"));
        insta::assert_debug_snapshot!(parse("class Foo { let A = 1"));
    }

    #[test]
    fn value() {
        insta::assert_debug_snapshot!(parse("class Foo<string A = \"hoge\" # \"fuga\">;"));
    }

    #[test]
    fn inner_value() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A = Hoge.Fuga, bits<2> B = Hoge{0...1}, list<int> C = Hoge[0...1]>;"
        ));
    }

    #[test]
    fn simple_value() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A = 1, string B = \"hoge\", bit D = false, int E = ?, bits<2> F = {0, 1}, list<int> G = [1, 2], dag H = (add A:$hoge), int I = A, int J = !add(A, B), int K = !cond(false: 1, true: 2)> {
                code C = [{ true }];
            }"
        ));
    }
}
