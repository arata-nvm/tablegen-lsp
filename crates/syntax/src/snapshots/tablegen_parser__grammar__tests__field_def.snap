---
source: tablegen-parser/src/grammar.rs
expression: "parse(\"class Foo { int\")"
---
(
    Root@0..15
      StatementList@0..15
        Class@0..15
          ClassKw@0..5 "class"
          Whitespace@5..6 " "
          Identifier@6..10
            Id@6..9 "Foo"
            Whitespace@9..10 " "
          RecordBody@10..15
            ParentClassList@10..10
            Body@10..15
              LBrace@10..11 "{"
              Whitespace@11..12 " "
              FieldDef@12..15
                IntType@12..15
                  Int@12..15 "int"
                Identifier@15..15
    ,
    [
        TableGenError {
            range: 15..15,
            message: "expected identifier in declaration",
        },
    ],
)
